src/SAWScript/Proof.hs

```diff
@@ -35,77 +407,476 @@ data ProofGoal =
+           , showTerm p
+           ]
+
+    checkTheorem :: Set Term -> Theorem -> IO ()
+    checkTheorem hyps (LocalAssumption p) =
+       unless (Set.member (unProp p) hyps) $ fail $ unlines
+          [ "Attempt to reference a local hypothesis that is not in scope"
+          , showTerm (unProp p)
+          ]
+    checkTheorem _hyps Theorem{} = return ()
+
+    check :: Set Term -> Evidence -> Prop -> IO ()
+    check hyps e p@(Prop ptm) = case e of
+      ProofTerm tm ->
+        do ty <- scTypeCheckError sc tm
+           ok <- scConvertible sc False ptm ty
+           unless ok $ fail $ unlines
+               [ "Proof term does not prove the required proposition"
+               , showTerm ptm
+               , showTerm tm
+               ]
+
+      LocalAssumptionEvidence (Prop l) ->
+        unless (Set.member l hyps) $ fail $ unlines
+             [ "Illegal use of local hypothesis"
+             , showTerm l
+             ]
+
+      SolverEvidence _stats (Prop p') ->
+        do ok <- scConvertible sc False ptm p'
+           unless ok $ fail $ unlines
+               [ "Solver proof does not prove the required proposition"
+               , showTerm ptm
+               , showTerm p'
+               ]
+
+      Admitted (Prop p') ->
+        do ok <- scConvertible sc False ptm p'
+           unless ok $ fail $ unlines
+               [ "Admitted proof does not match the required proposition"
+               , showTerm ptm
+               , showTerm p'
+               ]
+
+      QuickcheckEvidence _n (Prop p') ->
+        do ok <- scConvertible sc False ptm p'
+           unless ok $ fail $ unlines
+               [ "Quickcheck evidence does not match the required proposition"
+               , showTerm ptm
+               , showTerm p'
+               ]
+
+      TrivialEvidence ->
+        unless (propIsTrivial p) $ fail $ unlines
+            [ "Proposition is not trivial"
+            , showTerm ptm
+            ]
+
+      SplitEvidence e1 e2 ->
+        splitProp sc p >>= \case
+          Nothing -> fail $ unlines
+                       [ "Split evidence does not apply to non-conjunction prop"
+                       , showTerm ptm
+                       ]
+          Just (p1,p2) ->
+            do check hyps e1 p1
+               check hyps e2 p2
+
+      ApplyEvidence thm es ->
+        do checkTheorem hyps thm
+           p' <- checkApply hyps (thmProp thm) es
+           ok <- scConvertible sc False ptm p'
+           unless ok $ fail $ unlines
+               [ "Apply evidence does not match the required proposition"
+               , showTerm ptm
+               , showTerm p'
+               ]
+
+      CutEvidence thm e' ->
+        do checkTheorem hyps thm
+           p' <- scFun sc (unProp (thmProp thm)) ptm
+           check hyps e' (Prop p')
+
+      UnfoldEvidence vars e' ->
+        do p' <- unfoldProp sc vars p
+           check hyps e' p'
+
+      RewriteEvidence ss e' ->
```

Comment:
Note: this rule provides a bit of a soundness loophole.  Local assumptions can be added to simpsets, and we currently don't have any way to track their provenance, so they might be used out of their scope.  To fix this we will need to add some additional bookeeping to the `Simpset` type, which might be a bit awkward.